### PR TITLE
Issue #6946: add suppression filters to google and sun configs

### PR DESCRIPTION
**Pull request by romani**
Issue #6946

Examples of execution by CLI:
```
java -Dorg.checkstyle.google.suppressionxpathfilter.config="/var/tmp/xpath-suppressions.xml" \
  -jar /var/tmp/checkstyle-8.28-all.jar \
  -c google_checks.xml \
  Test.java
Starting audit...
Audit done.
```
```
java -Dorg.checkstyle.google.suppressionfilter.config="/var/tmp/suppressions.xml" \
  -jar /var/tmp/checkstyle-8.28-all.jar \
  -c google_checks.xml \
  Test.java
Starting audit...
Audit done.
```

Tested at https://github.com/sevntu-checkstyle/checkstyle-samples/pull/23

### CODE REVIEW

**Comment by romani** [comment]
@rnveach , @pbludov , @strkkk , please review this change.

I still thinking on how to cover it by tests 
and probably do some examples to show users how to use this. 
I you have good ideas please share.

File: src/xdocs/config_filters.xml [hunk 1]
```diff
@@ -1391,9 +1399,17 @@ CLASS_DEF -> CLASS_DEF [1:0]
...
+              Google Style</a>
+          </li>
+          <li>
+            <a href="https://github.com/search?q=path%3Asrc%2Fmain%2Fresources+filename%3Asun_checks.xml+repo%3Acheckstyle%2Fcheckstyle+SuppressionXpathFilter">
+              Sun Style</a>
+          </li>
           <li>
             <a href="https://github.com/search?q=path%3Aconfig+filename%3Acheckstyle_checks.xml+repo%3Acheckstyle%2Fcheckstyle+SuppressionXpathFilter">
-              Checkstyle Style</a>
+            Checkstyle Style</a>
```
Comment (rnveach) [inline src/xdocs/config_filters.xml]: Why was this extra spacing removed but it looks like you have it for Sun and Google?

Comment (romani) [reply src/xdocs/config_filters.xml]: fixed.

**Comment by romani** [comment]
Yes I assigned or to me as sign that I need to do something on it

**Comment by romani** [comment]
@rnveach , ready for review.

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java [hunk 2]
```diff
@@ -417,6 +420,30 @@ public void testExistingTargetFileWithViolations(@SysErr Capturable systemErr,
         assertEquals("", systemErr.getCapturedData(), "Unexpected system error log");
     }
 
+    @Test
+    public void testViolationsByGoogleAndXpathSuppressions(@SysErr Capturable systemErr,
+            @SysOut Capturable systemOut) throws Exception {
+        System.setProperty("org.checkstyle.google.suppressionxpathfilter.config",
+                "src/test/resources/com/puppycrawl/tools/checkstyle/main/"
+                        + "InputMainViolationsForGoogleXpathSuppressions.xml");
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: Why can't `getPath` be used instead?

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java [hunk 3]
```diff
@@ -417,6 +420,30 @@ public void testExistingTargetFileWithViolations(@SysErr Capturable systemErr,
...
+        assertThat("Unexpected output log", systemOut.getCapturedData(), is(noViolationsOutput));
+        assertThat("Unexpected system error log", systemErr.getCapturedData(), is(""));
+    }
+
+    @Test
+    public void testViolationsByGoogleAndSuppressions(@SysErr Capturable systemErr,
+            @SysOut Capturable systemOut) throws Exception {
+        System.setProperty("org.checkstyle.google.suppressionfilter.config",
+                "src/test/resources/com/puppycrawl/tools/checkstyle/main/"
+                        + "InputMainViolationsForGoogleSuppressions.xml");
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: same

File: src/main/resources/google_checks.xml [hunk 4]
```diff
@@ -11,6 +11,7 @@
     http://checkstyle.org (or in your downloaded distribution).
 
     To completely disable a check, just comment it out or delete it from the file.
+    To suppress certain violations please review suppression filters.
```
Comment (rnveach) [inline src/main/resources/google_checks.xml]: As of right now, there is no way for users to know how to use suppression in custom checks without looking at the file and seeing the property name to be used. What do you think of expanding the google/sun reports to specify how to use the suppression. I would think a short description, specifying the property names and pointing to the other filter documentation would be enough.

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java (hunk 2 above)
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: fixed

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java (hunk 3 above)
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: fixed

File: src/main/resources/google_checks.xml (hunk 4 above)
Comment (romani) [reply src/main/resources/google_checks.xml]: lets me finish with examples of usage .... and based on this we can think what words are better to use....

Comment (romani) [reply src/main/resources/google_checks.xml]: here are examples in maven - https://github.com/sevntu-checkstyle/checkstyle-samples/pull/23
looks good
executed as (8.28-SNAPSHOT is based on this branch code):
```
✔ ~/java/github/sevntu-checkstyle/checkstyle-samples [google-samples-suppression L|✔] 
$ mvn -f pom-google.xml clean checkstyle:checkstyle -Dcheckstyle.version=8.28-SNAPSHOT

$ mvn -f pom-google-custom-suppression.xml clean checkstyle:checkstyle -Dcheckstyle.version=8.28-SNAPSHOT

```

**Comment by romani** [comment]
some first step for documentation:
image
if you know how to improve please suggest.

File: src/main/resources/google_checks.xml (hunk 4 above)
Comment (romani) [reply src/main/resources/google_checks.xml]: done.

**Comment by romani** [comment]
@rdiachenko , I tested you use case , see description of PR for link.
Can you recheck that this works on your non-open sources ?

**Comment by rnveach** [comment]
> I tried code tag ... but does not looks very different from ordinary text.

If you want the text to look different we can always override with CSS.

File: src/xdocs/sun_style.xml [hunk 5]
```diff
@@ -997,6 +997,30 @@
...
+            SuppressionFilter</a>
+          and
+          <a href="config_filters.html#SuppressionXpathFilter">
+            SuppressionXpathFilter</a>.
+          Location of config file for <i>SuppressionFilter</i> can be defined by system property
+          <i>org.checkstyle.google.suppressionfilter.config</i> (default value is
+          <i>checkstyle-suppressions.xml</i>).
+          Location of config file for <i>SuppressionXpathFilter</i>
+          can be defined by system property
+          <i>org.checkstyle.google.suppressionxpathfilter.config</i> (default value is
```
Comment (rnveach) [inline src/xdocs/sun_style.xml]: Shouldn't this be sun for sun config?

File: src/xdocs/sun_style.xml [hunk 6]
```diff
@@ -997,6 +997,30 @@
...
+      <subsection name="Suppressions" id="Sun_Suppressions">
+        <p>
+          It is possible to suppress some violations by embeded filters
+          <a href="config_filters.html#SuppressionFilter">
+            SuppressionFilter</a>
+          and
+          <a href="config_filters.html#SuppressionXpathFilter">
+            SuppressionXpathFilter</a>.
+          Location of config file for <i>SuppressionFilter</i> can be defined by system property
+          <i>org.checkstyle.google.suppressionfilter.config</i> (default value is
```
Comment (rnveach) [inline src/xdocs/sun_style.xml]: same.

File: src/xdocs/sun_style.xml (hunk 5 above)
Comment (romani) [reply src/xdocs/sun_style.xml]: fixed.

File: src/xdocs/sun_style.xml (hunk 6 above)
Comment (romani) [reply src/xdocs/sun_style.xml]: fixed.

**Comment by rdiachenko** [comment]
> @rdiachenko , I tested your use case , see description of PR for link.
> 
> Can you recheck that this works on your non-open sources ?

@romani , works fine. Thanks a lot!

**Pull request merged**